src/raid.cpp

```diff
@@ -0,0 +1,925 @@
+    {
+        if (unusedRaidConnection < RAIDPARTS)
+        {
+            LOG_warn << "5 connection cloudraid shutting down connection " << errorConnectionNum << " due to error, and starting " << unusedRaidConnection << " instead";
+
+            // start up the old unused connection, and cancel this one.  Other connections all have real data since we were already in 5 connection mode
+            clearOwningFilePieces(raidinputparts[unusedRaidConnection]);
+            clearOwningFilePieces(raidinputparts[errorConnectionNum]);
+            raidrequestpartpos[unusedRaidConnection] = raidpartspos;
+            raidrequestpartpos[errorConnectionNum] = raidpartspos;
+        }
+        else
+        {
+            LOG_warn << "6 connection cloudraid shutting down connection " << errorConnectionNum << " due to error";
+            clearOwningFilePieces(raidinputparts[errorConnectionNum]);
+            raidrequestpartpos[errorConnectionNum] = raidpartspos;
+        }
+
+        unusedRaidConnection = errorConnectionNum;
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool RaidBufferManager::connectionRaidPeersAreAllPaused(unsigned slowConnection)
+{
+    if (!isRaid())
+    {
+        return false;
+    }
+
+    // see if one connection is stalled or running much slower than the others, in which case try the other 5 instead
+    for (unsigned j = RAIDPARTS; j--; )
+    {
+        if (j != slowConnection && j != unusedRaidConnection && !connectionPaused[j])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool RaidBufferManager::detectSlowestRaidConnection(unsigned thisConnection, unsigned& slowestConnection)
+{
+    if (isRaid() && unusedRaidConnection == RAIDPARTS)
+    {
+        connectionStarted[thisConnection] = true;
+        int count = 0;
+        for (unsigned j = RAIDPARTS; j--; )
+        {
+            if (!connectionStarted[j])
+            {
+                slowestConnection = j;
+                ++count;
+            }
+        }
+        if (count == 1)
+        {
+            unusedRaidConnection = slowestConnection;
+            raidrequestpartpos[unusedRaidConnection] = raidpartspos;
+            return true;
+        }
+    }
+    return false;
+}
+
+TransferBufferManager::TransferBufferManager()
+    : transfer(NULL)
+{
+}
+
+void TransferBufferManager::setIsRaid(Transfer* t, std::vector<std::string>& tempUrls, m_off_t resumepos, m_off_t maxRequestSize)
+{
+    RaidBufferManager::setIsRaid(tempUrls, resumepos, t->size, t->size, maxRequestSize);
+
+    transfer = t;
+}
+
+m_off_t& TransferBufferManager::transferPos(unsigned connectionNum)
+{
+    return isRaid() ? RaidBufferManager::transferPos(connectionNum) : transfer->pos;
+}
+
+m_off_t TransferBufferManager::nextTransferPos()
+{
+    assert(!isRaid());
+    chunkmac_map& chunkmacs = transfer->chunkmacs;
+    while (chunkmacs.find(ChunkedHash::chunkfloor(transfer->pos)) != chunkmacs.end())
```

Comment:
I don't know if the `chunkmacs` map may contain a lot of items, but since the item resulting from the `find()` is used inside the block perhaps is a good idea to declare an iterator and use it afterwords, avoiding 2 searches instead of one. I'm adding some comments, so I'll change it unless you prefer to keep it as it is currently ;)

Reply:
Hi Sergio, this is actually code that existed before (in transferslot) but I've moved it here as part of the abstraction.   As transfer->pos is being updated in the loop, and that's what find() is searching for, I think it's best left as is in this case (though, generally that is a good idea of course!)  thanks

Reply:
I've added a commit (not pushed yet) like this:
```
    chunkmac_map::const_iterator it = chunkmacs.find(ChunkedHash::chunkfloor(transfer->pos));
    while (it != chunkmacs.end())
    {
        if (it->second.finished)
        {
            transfer->pos = ChunkedHash::chunkceil(transfer->pos);
        }
        else
        {
            transfer->pos += it->second.offset;
            break;
        }
        it = chunkmacs.find(ChunkedHash::chunkfloor(transfer->pos));
    }```
But I can revert it, if you feel it's better to not change it :smiley:

Reply:
Ah, I see what you mean now - yes that looks like a good improvement, thanks